lib/pure/asyncdispatch.nim

```diff
@@ -355,6 +355,28 @@ proc `or`*[T, Y](fut1: Future[T], fut2: Future[Y]): Future[void] =
   fut2.callback = cb
   return retFuture
 
+proc all*[A](futs: seq[Future[A]]): Future[seq[A]] =
+  ## Returns a future which will complete once all futures in ``futs``
+  ## complete.
+  ##
+  ## The resulting future will hold the values of all awaited futures,
+  ## in the order they are passed.
+
+  var
+    retFuture = newFuture[seq[A]]("asyncdispatch.all")
+    retValues = newSeq[A](len(futs))
+    completedFutures = 0
+
+  for i, fut in futs:
+    fut.callback = proc(f: Future[A]) =
+      retValues[i] = f.read()
```

Comment:
The `i` value here will always be the last one. 

A suggestion:

```nim
type 
  AllContext[A] = ref object
    i: int
    completedFutures: int
    fut: Future[A]
    retValues: seq[A]
    length: int
    retFuture: Future[seq[A]]

proc allCallback[A](x: AllContext[A]) = 
  var index = x.i
  x.fut.callback = proc(f: Future[A]) = 
    x.retValues[index] = f.read()
    inc(x.completedFutures)
    if x.completedFutures == x.length:
      x.retFuture.complete(x.retValues)

proc all*[A](futs: varargs[Future[A]]): Future[seq[A]] =
  var
    retFuture = newFuture[seq[A]]("asyncdispatch.all")
    retValues = newSeq[A](len(futs))
    context = new(AllContext[A])
  context.completedFutures = 0
  context.retValues = retValues
  context.length = len(futs)
  context.retFuture = retFuture

  for i, fut in futs:
    context.i = i
    context.fut =fut
    allCallback[A](context)

  return retFuture
```

Test example:

```nim
import asyncdispatch

proc fut1(name: string): Future[string] {.async.} = 
  await sleepAsync(1000)
  result = "hello " & name

proc main() {.async.} =
  var f1 = fut1("A")
  var f2 = fut1("B")
  var ret = await all(f1, f2)
  echo ret

waitFor main()
```